developer/src/kmc-keyboard-info/src/index.ts

```diff
@@ -0,0 +1,412 @@
+ * compiled files to produce a comprehensive .keyboard_info file.
+ */
+
+import { minKeymanVersion } from "./min-keyman-version.js";
+import { KeyboardInfoFile, KeyboardInfoFileIncludes, KeyboardInfoFileLanguage, KeyboardInfoFilePlatform } from "./keyboard-info-file.js";
+import { KeymanFileTypes, CompilerCallbacks, KmpJsonFile, KmxFileReader, KMX, KeymanTargets } from "@keymanapp/common-types";
+import { KeyboardInfoCompilerMessages } from "./messages.js";
+import langtags from "./imports/langtags.js";
+
+const regionNames = new Intl.DisplayNames(['en'], { type: "region" });
+const scriptNames = new Intl.DisplayNames(['en'], { type: "script" });
+const langtagsByTag = {};
+
+/**
+ * Build a dictionary of language tags from langtags.json
+ */
+
+function init(): void {
+  if(langtagsByTag['en']) {
+    // Already initialized, we can reasonably assume that 'en' will always be in
+    // langtags.json.
+    return;
+  }
+
+  for(const tag of langtags) {
+    langtagsByTag[tag.tag] = tag;
+    langtagsByTag[tag.full] = tag;
+    if(tag.tags) {
+      for(const t of tag.tags) {
+        langtagsByTag[t] = tag;
+      }
+    }
+  }
+}
+
+/* c8 ignore start */
+export class KeyboardInfoOptions {
+  /** The identifier for the keyboard */
+  keyboard_id: string;
+
+  /** The data from the .kps file, transformed to kmp.json */
+  kmpJsonData: KmpJsonFile.KmpJsonFile;
+
+  /** The path in the keymanapp/keyboards repo where this keyboard may be found (optional) */
+  sourcePath?: string;
+
+  /** The full URL to the keyboard help, starting with https://help.keyman.com/keyboard/ (optional) */
+  helpLink?: string;
+
+  /** The compiled keyboard filename and relative path (.js only) */
+  keyboardFileNameJs?: string;
+
+  /** The compiled package filename and relative path (.kmp) */
+  kmpFileName: string;
+
+  /** The source package filename and relative path (.kps) */
+  kpsFileName: string;
+};
+/* c8 ignore stop */
+
+export class KeyboardInfoCompiler {
+  constructor(private callbacks: CompilerCallbacks) {
+    init();
+  }
+
+  /**
+   * Merges source .keyboard_info file with metadata from the keyboard and package source file.
+   * This function is intended for use within the keyboards repository. While many of the
+   * parameters could be deduced from each other, they are specified here to reduce the
+   * number of places the filenames are constructed.
+   * For full documentation, see:
+   * https://help.keyman.com/developer/cloud/keyboard_info/
+   *
+   * @param sourceKeyboardInfoFileName  Path for the source .keyboard_info file
+   * @param options                     Details on files from which to extract additional metadata
+   */
+  public writeMergedKeyboardInfoFile(
+    sourceKeyboardInfoFileName: string,
+    options: KeyboardInfoOptions
+  ): Uint8Array {
+    let keyboard_info: KeyboardInfoFile = null;
+
+    if(this.callbacks.fs.existsSync(sourceKeyboardInfoFileName)) {
```

Comment:
So, since I'm not 100% up-to-date on the design here... we have a _source_ `.keyboard_info` file that gets extended with additional metadata for a final, built output version?  Or is something else going on here?

Just trying to make sure my mental model is reasonably accurate, as that might affect the quality of this review.

Reply:
> So, since I'm not 100% up-to-date on the design here... we have a _source_ `.keyboard_info` file that gets extended with additional metadata for a final, built output version?

Correct. In hindsight, using the same extension was a recipe for pain. But hindsight 100% yada yada.

So the final output is in the build/ folder and is generated from both the source .keyboard_info and metadata found in the .kps and .kmn (.kmx/.js) files. A future update will make it possible to generate the final .keyboard_info without having a source .keyboard_info at all, and _eventually_ I will eliminate source .keyboard_info entirely and have it generated during the .kmp build from the files in the project.